apps/mobile/src/utils/notifications/cleanup.ts

```diff
@@ -0,0 +1,166 @@
+        if (attempt < maxRetries) {
+          const delay = baseDelay * Math.pow(2, attempt - 1) // Exponential backoff
+          Logger.info(`Rate limited, retrying in ${delay}ms (attempt ${attempt}/${maxRetries})`)
+          await new Promise((resolve) => setTimeout(resolve, delay))
+          continue
+        }
+      }
+      throw error
+    }
+  }
+  throw new Error('Should not reach here')
+}
+
+/**
+ * Gets all safes affected by removing a delegate (pure function)
+ */
+export const getAffectedSafes = (
+  ownerAddress: Address,
+  allSafes: RootState['safes'],
+  allChains: { chainId: string }[],
+  safeSubscriptions: RootState['safeSubscriptions'],
+): SafeNotificationInfo[] => {
+  const affectedSafes: SafeNotificationInfo[] = []
+
+  Object.entries(allSafes).forEach(([safeAddress, chainDeployments]) => {
+    // Check if this owner is part of this safe
+    const isOwner = Object.values(chainDeployments).some((deployment) =>
+      deployment.owners.some((owner) => owner.value === ownerAddress),
+    )
+
+    if (isOwner) {
+      // Get chains where this safe is subscribed to notifications
+      const subscribedChains = allChains
+        .map((chain) => chain.chainId)
+        .filter((chainId) => {
+          const subscriptionStatus = safeSubscriptions[safeAddress]?.[chainId]
+          return subscriptionStatus === true
+        })
+
+      if (subscribedChains.length > 0) {
+        affectedSafes.push({
+          address: safeAddress,
+          chainIds: subscribedChains,
+        })
+      }
+    }
+  })
+
+  return affectedSafes
+}
+
+/**
+ * Checks if a safe has other delegates besides the one being removed (pure function)
+ */
+export const hasOtherDelegates = (safeAddress: Address, excludeDelegateAddress: Address, state: RootState): boolean => {
+  const allSafeDelegates = selectAllDelegatesForSafeOwners(state, safeAddress)
+  return allSafeDelegates.some((delegate) => delegate.delegateAddress !== excludeDelegateAddress)
+}
+
+/**
+ * Creates subscription data for API calls
+ */
+export const createSubscriptionData = async (
+  safeAddress: string,
+  chainIds: string[],
+  deviceUuid: string,
+  delegateAddress?: string,
+) => {
+  return chainIds.map((chainId) => ({
+    chainId,
+    deviceUuid,
+    safeAddress,
+    ...(delegateAddress && { signerAddress: delegateAddress }),
+  }))
+}
+
+/**
+ * Clears authentication cookies before making non-authenticated API calls.
+ * This is a workaround for the React Native bug where credentials: 'omit' is ignored
+ * on Android and cookies are always sent.
+ *
+ * We do inline import to avoid circular dependency.
+ *
+ * @see https://github.com/facebook/react-native/issues/12956
+ */
+export const clearAuthBeforeUnauthenticatedCall = async (): Promise<void> => {
+  try {
+    const { cgwApi: authApi } = await import('@safe-global/store/gateway/AUTO_GENERATED/auth')
+    const { getStore } = await import('@/src/store/utils/singletonStore')
```

Comment:
Is there a reason we are importing this inside the function and not outside?

Reply:
yes, it's in the function description - circular dependency errors

/**
 * Clears authentication cookies before making non-authenticated API calls.
 * This is a workaround for the React Native bug where credentials: 'omit' is ignored
 * on Android and cookies are always sent.
 *
 * We do inline import to avoid circular dependency.
 *
 * @see https://github.com/facebook/react-native/issues/12956
 */